### PR TITLE
Native packager

### PR DESCRIPTION
**Pull request by tdroxler**
Resolves: https://github.com/alephium/ralph-lsp/issues/165 

This PR introduce [sbt-native-packager](https://github.com/sbt/sbt-native-packager) which allows to : 

Build native packages for different systems
  * Universal `zip`,`tar.gz`, `xz` archives
  * `deb` and `rpm` packages for Debian/RHEL based systems
  * `dmg` for macOS
  * `msi` for Windows
  * `docker` images
  * `graalvm` native images

I started with the `zip` archive, but we could later choose to add some others. In the archive, you have two folders: `bin` and `lib`,

in `bin` you have the `ralph-lsp` executable for Linux and MacOs, as well a `ralph-lsp.bat` for Windows.

Using the `sbt stage` command allows to create the `bin/lib` folders locally at `lsp-server/target/universal/stage` and is convenient for local development.
Adding `lsp-server/target/universal/stage/bin/` to your `PATH` gives you then `ralph-lsp` executable from anywhere and we don't need to run `java -jar ralph-lsp.jar`

`scaladoc` was failing in multiple places, see #172

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -92,6 +94,7 @@ lazy val copyJARToVSCode =
 lazy val `lsp-server` =
   project
     .settings(
+      name := "ralph-lsp",
```
Comment (tdroxler) [inline build.sbt:97]: moving the name here because of this: https://github.com/sbt/sbt-native-packager/issues/1530

**Comment by tdroxler** [comment]
@simerplaha could you try the `sbt universal:packageOsxDmg` command and then try `ralph-lsp.dmg`?

**Comment by simerplaha** [comment]
> @simerplaha could you try the `sbt universal:packageOsxDmg` command and then try `ralph-lsp.dmg`?

Executing the generated `ralph-lsp-0.1.0-SNAPSHOT.dmg` mounts a volume. Running the executable file `ralph-lsp` from the mounted volume starts the server.

**Comment by simerplaha** [comment]
Hey Thomas, just a quick question for my knowledge. I'm trying to understand the motivation here. Please help me understand.

> Adding `lsp-server/target/universal/stage/bin/` to your `PATH` gives you then `ralph-lsp` executable from anywhere and we don't need to run `java -jar ralph-lsp.jar`

Does `we don't need to run java -jar ralph-lsp.jar` mean that executing `java -jar ralph-lsp.jar` is a problem for something/somewhere?

> Every lsp server I use have an executable and it has to be in the PATH otherwise nvim can't launch it, applies for `metals`, `lua-language-server`  etc (those are the executable). 

And from your [comment above](https://github.com/alephium/ralph-lsp/issues/165#issue-2210785314), I'm understanding  that NVIM requires a native executable file and our current setup of using `java -jar ralph-lsp.jar` does not work for NVIM and therefore is a problem for implementing issue #164?

File: build.sbt (hunk 1 above)
Comment (tdroxler) [reply build.sbt:97]: :+1: good idea

**Comment by tdroxler** [comment]
>And from your https://github.com/alephium/ralph-lsp/issues/165#issue-2210785314, I'm understanding that NVIM requires a native executable file and our current setup of using java -jar ralph-lsp.jar does not work for NVIM and therefore is a problem for implementing issue https://github.com/alephium/ralph-lsp/issues/164?

Running `java -jar ralph-lsp.jar` works, but you need to know and give the full path of  `ralph-lsp.jar`, while the more regular way to launch any program (really any like firefox, metals, intelij ...), you have an executable that is somewhere in your `PATH` and you just call that program to start it. 
On windows you usually have a boring installation phase, but what it does is mostly copying the program into `Program Files`
On MacOs you probably know better, but to install something you just drag and drop the dmg to the `Application` folder no?

For our vs-code extension it's currently fine as we pack the server inside the client.

For nvim, when configuring the client you pass from:
```lua
   vim.lsp.start({
     name = 'ralph-lsp',
      cmd = {'java', '-jar', '/home/thomas/dev/ralph-lsp/lsp-server/target/scala-2.13/ralph-lsp.jar'}
   })
```
to
```lua
   vim.lsp.start({
     name = 'ralph-lsp',
      cmd = {'ralph-lsp'}
   })
```
When you can have an executable available in your PATH.

**Comment by simerplaha** [comment]
Thank you. So NVIM can launch `java -jar ralph-lsp.jar` and it is not a problem. Also, the plugin for NVIM in #164 can be implemented without introducing `sbt-native-packager`.

I'm still unable to figure out the problem, that we currently have, which is being solved by introducing `sbt-native-packager` 🤔.

Is it a convenience thing because `cmd = {'ralph-lsp'}` is much shorter than `cmd = {'java', '-jar', '/home/thomas/dev/ralph-lsp/lsp-server/target/scala-2.13/ralph-lsp.jar'}`?

Should we consider delaying the introduction of `sbt-native-packager` as dependency and the extra build code it brings until we actually run into a problem that needs it?

**Comment by tdroxler** [comment]
It's not convenient because it's shorter, it's convenient because user don't need to set the path to the jar when installing a plugin. Even for the vs-code plugin, we wouldn't need to pack the jar into the plugin, just need to have `ralph-lsp` installed. That will be the case for emacs or any other editor.

[metals](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/metals.lua#L5), [rust](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/rust_analyzer.lua#L45), [typescript](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/tsserver.lua#L6), [c#](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/csharp_ls.lua#L5), [eclipse-java-lsp](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/jdtls.lua#L93-L100), [lua](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/lua_ls.lua#L15), [solidity](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/solidity.lua#L5), [swift](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/sourcekit.lua#L5), [python](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/pyright.lua#L45), [go](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/gopls.lua#L7) and I could continue, they all providing an executable.

I found one that has to provide the jar path, [drools](https://github.com/neovim/nvim-lspconfig/blob/e25c4cdecd3d58c0deccce0f372426c8c480bcce/lua/lspconfig/server_configurations/drools_lsp.lua#L58) 

>Should we consider delaying the introduction of sbt-native-packager as dependency and the extra build code it brings until we actually run into a problem that needs it?

For scalafmt, even tho I don't agree, I could understand your concern, but here I just don't see the problem. It's a build/CI thing and we'll still continue to publish the jar file too

**Comment by simerplaha** [comment]
Let's go your way 👍🏼.

**Comment by tdroxler** [comment]
@simerplaha I also took the opportunity to use `matrix` in our workflow so it's cleaner and I also added testing MacOS. 
If we want we could also add a matrix for java versions, for example if we would like to test on java 11 and 17

**Review by simerplaha — APPROVED** [review]
👍🏼

**Pull request merged**